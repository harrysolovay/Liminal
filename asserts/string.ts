import { assert } from "../util/mod.ts"

export function minLength(value: string, minLength: number) {
  assert(value.length >= minLength, `Length must be gte ${minLength} chars.`)
}

export function maxLength(value: string, maxLength: number) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} chars.`)
}
