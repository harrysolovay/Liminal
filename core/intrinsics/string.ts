import type { Rune } from "../Rune.ts"
import { Type } from "../Type.ts"

export const string: Rune<string, never> = Type("string", () => string)
