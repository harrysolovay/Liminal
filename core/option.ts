import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"

export function option<X extends Type>(Some: X): Type<X["T"] | undefined, {}, X["P"]> {
  return declare({
    name: "option",
    source: {
      factory: option,
      args: { Some },
    },
    subschema: (visit) => ({
      discriminator: "type",
      anyOf: [{ type: "null" }, visit(Some)],
    }),
    output: (f) =>
      f<X["T"] | null>({
        visitor: (value, visit) => value === null ? undefined : visit(value, Some, "Some"),
      }),
  })
}
