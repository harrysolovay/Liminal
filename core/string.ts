import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare<string>()({
  name: "string",
  source: {
    getType: () => string,
  },
  subschema: () => ({ type: "string" }),
  visitor: (value) => value,
  assertRefinementsValid: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
    )
  },
  assertRefinements: {
    minLength: (value, minLength) => assert(value.length >= minLength),
    maxLength: (value, maxLength) => assert(value.length <= maxLength),
  },
})
