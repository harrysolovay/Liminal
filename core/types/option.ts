import { type AnyType, Type } from "../Type.ts"

export function option<X extends AnyType>(Some: X): Type<X["T"] | undefined, X["P"]> {
  return Type({
    name: "option",
    source: {
      factory: option,
      args: [Some],
    },
  })
}
