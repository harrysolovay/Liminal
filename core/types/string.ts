import { Type } from "../Type.ts"

export const string: Type<string> = Type<string>({
  name: "string",
  source: {
    getType: () => string,
  },
})
