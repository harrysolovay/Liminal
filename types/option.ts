import { declare, type Type } from "../core/mod.ts"

export function option<X extends Type>(Some: X): Type<X["T"] | undefined, {}, X["P"]> {
  return declare({
    name: "option",
    source: {
      factory: option,
      args: { Some },
    },
    subschema: (visit) => ({
      discriminator: "type",
      anyOf: [{ type: "null" }, visit(Some)],
    }),
    output: (f) =>
      f<X["T"] | null>({
        visitor: (value, ctx) => {
          return value === null ? undefined : ctx.visit({
            value,
            type: Some,
            junctions: {
              type: "Some",
            },
          })
        },
      }),
  })
}
