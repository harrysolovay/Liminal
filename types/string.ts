import { declare, type Type } from "../core/mod.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare({
  name: "string",
  source: {
    getType: () => string,
  },
  refinements: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      "`minLength` cannot be greater than `maxLength`.",
    )
    return {
      minLength: `Length must be gte ${minLength} chars.`,
      maxLength: `Length must be lte ${maxLength} chars.`,
    }
  },
  subschema: () => ({ type: "string" }),
  output: (f) =>
    f({
      refinementPredicates: {
        minLength: (value, minLength) => value.length >= minLength,
        maxLength: (value, maxLength) => value.length <= maxLength,
      },
    }),
})
