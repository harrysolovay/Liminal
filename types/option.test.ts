import { assertTypeSnapshot } from "testing"
import { option } from "./option.ts"
import { string } from "./string.ts"

Deno.test("option", async (t) => {
  await assertTypeSnapshot(t, option(string))
})
