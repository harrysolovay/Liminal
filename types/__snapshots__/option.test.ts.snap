export const snapshot = {};

snapshot[`option 1`] = `
{
  anyOf: [
    {
      type: "null",
    },
    {
      type: "string",
    },
  ],
  discriminator: "type",
}
`;

snapshot[`option 2`] = `
{
  anyOf: [
    {
      type: "null",
    },
    {
      type: "string",
    },
  ],
  description: "One.",
  discriminator: "type",
}
`;

snapshot[`option 3`] = `
{
  anyOf: [
    {
      type: "null",
    },
    {
      type: "string",
    },
  ],
  description: "One. Two.",
  discriminator: "type",
}
`;
